### PR TITLE
Add empty

### PR DESCRIPTION
**Pull request by jakirkham**
Adds a function for creating an empty generator. Very handy for cases where a generator is required, but we don't want to actually yield anything.

### CODE REVIEW

File: toolz/itertoolz.py [hunk 1]
```diff
@@ -9,14 +9,27 @@
...
+__all__ = ('empty', 'remove', 'accumulate', 'groupby', 'merge_sorted',
+           'interleave', 'unique', 'isiterable', 'isdistinct', 'take',
+           'drop', 'take_nth', 'first', 'second', 'nth', 'last', 'get',
+           'concat', 'concatv', 'mapcat', 'cons', 'interpose',
+           'frequencies', 'reduceby', 'iterate', 'sliding_window',
+           'partition', 'partition_all', 'count', 'pluck',
            'join', 'tail', 'diff', 'topk', 'peek', 'random_sample')
 
 
+def empty():
```
Comment (llllllllll) [inline toolz/itertoolz.py]: a simple way to do this is `return iter(())`

Comment (jakirkham) [inline toolz/itertoolz.py]: Yeah that would work too. Was kind of wanting it to be a `generator` type. Though if there is a strong preference to change, it we can do that.

**Comment by jakirkham** [comment]
Weird. The tests pass locally and on CI, but somehow we are getting build failures. 😕

**Comment by llllllllll** [comment]
It is a coverage error (which is why the pypy tests pass). toolz enforces 100% line coverage for the code since the utilities are simple enough that hitting all routes shouldn't be too hard. It looks like line 30: (`yield _`) is never hit because the iterator is always empty.

**Comment by jakirkham** [comment]
Yep, ok, that makes sense. Have switched to `iter([])` as suggested, which shouldn't have this problem.

**Comment by jakirkham** [comment]
Alright, I think this is good now. Thanks for the tips.

File: toolz/itertoolz.py (hunk 1 above)
Comment (llllllllll) [inline toolz/itertoolz.py]: Does the type matter? Did you need the `send` and `throw` functions? I think that  extends the API from iterator to coroutine. If you are interested in coroutine utils you can look at `cotoolz`, though I haven't updated it to use the latest toolz. This provides an `emptycoroutine` object which has a send, next, and throw method which behave like an empty coroutine (generator)

Comment (jakirkham) [inline toolz/itertoolz.py]: Kind of, but not for any good reasons. I'm happy with how this PR has ended up. The functionality provided by this function/iterator is sufficient.

**Comment by mrocklin** [comment]
Just checking, is this something that you need or is it something that you think should be in toolz?

There is some non-trivial cost to expanding the API.  We should check that each addition is necessary.  Also, are you willing to add Cythonized versions of the recent PRs to cytoolz as well?

**Comment by jakirkham** [comment]
Both. I'm refactoring some of my code into smaller libraries. Some of it included iterator stuff. I thought it would be desirable to collaborate on an existing, well-established library instead of rolling my own. I may still need to do this short term, but I'd much rather move towards collaborating if at all possible. `toolz` seems like the ideal place for these additions IMHO.

So I haven't found a need to get these to go any faster, personally. However, I'm not opposed to adding them to `cytoolz` if that is important. Though I can't promise to spend any time on it in the near term.

**Comment by mrocklin** [comment]
I mostly mention `cytoolz` as a basic barrier to entry.  Someone does need to add these to cytoolz to maintain cross compatibility, even if they're just Python implementations.  Usually this falls to @eriknw but it'd be nice to spread around the workload.  

If we can't find someone to add these functions to cytoolz then we might want to think harder about whether or not we should include them.  I want to avoid toolz becoming a dumping ground for functions without strong champions.

**Comment by jakirkham** [comment]
Again I'm not against doing it long term. I just have to be honest and state that near term I'm trying to publish something. So adding things to `cytoolz` is not a priority.

**Comment by jakirkham** [comment]
Also it would be nice to know if these are things that `toolz` is likely to accept assuming working `cytoolz` implementations before sinking time into writing Cython implementations. If not, I would rather save us all the trouble.

**Comment by mrocklin** [comment]
I don't intend to make a decision here (I've been pretty absent when it comes to toolz development).  I suggest that others like @llllllllll and @eriknw who are more active these days take the lead on inclusion decisions.

**Comment by eriknw** [comment]
Hey!  I've been mostly-afk for the last several weeks (doing this from time to time is good for my overall well-being).  I'm generally happy to keep `cytoolz` up to date with `toolz` on my own, although I'm behind atm.  Any help is, of course, appreciated, but is not something I expect.

> I want to avoid toolz becoming a dumping ground for functions without strong champions.

I agree with this sentiment, although I also think we need to lower the barrier for accepting contributions for functions that do have strong champions as long as they adhere to the style and principles of `toolz`.  Limiting the API--we've tried hard to keep the number of functions in `toolz` low--keeps the library user-friendly and approachable for new users.  Limiting the API, however, frustrates contributors (not to mention we miss out on some perfectly reasonable functionality), and I would like to find a better way to balance this tension.  Perhaps we should no longer import everything to the `toolz` namespace.  This will give everybody more freedom to add functions to submodules, and the top `toolz` namespace will remain the curated list of the top functions for people to explore.

**Comment by llllllllll** [comment]
> Perhaps we should no longer import everything to the toolz namespace.

I really like the idea of having more pre-composed helpers but leaving the top level api simple atomic operations. This keeps things simple for discovery but once you are comfortable with the atomics you can dig into the more rich functionality.

That said, I think the current proposal for `empty` doesn't provide enough value over just writing `iter(())`. One way we could make it more helpful would be to make it a custom object that had a special `repr` which would aid debugging by showing `<empty iterator>` or something so users know what they are working with.

**Comment by mrocklin** [comment]
I'm generally -1 on anything that doesn't provide much functionality other than a pretty repr.

**Comment by jakirkham** [comment]
Alright, I don't want to bikeshed this trivial addition. So I think I'll just close it. If I'm missing something and you do want this, please let me know.

**Pull request closed without merging**